File: 2024/day11/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <map>
#include <cmath>

using namespace std;

// Memoization cache: (value, blinks) -> count
map<pair<long long, int>, long long> memo;

// Count digits in a number
int count_digits(long long n) {
    if (n == 0) return 1;
    int count = 0;
    while (n > 0) {
        count++;
        n /= 10;
    }
    return count;
}

// Split a number with even digits into left and right halves
pair<long long, long long> split_number(long long value, int num_digits) {
    int mid = num_digits / 2;
    long long divisor = 1;
    for (int i = 0; i < mid; i++) {
        divisor *= 10;
    }
    long long left = value / divisor;
    long long right = value % divisor;
    return {left, right};
}

// Count how many stones result from a single stone after N blinks
long long count_stones(long long value, int blinks) {
    // Base case
    if (blinks == 0) {
        return 1;
    }

    // Check memoization cache
    auto key = make_pair(value, blinks);
    if (memo.find(key) != memo.end()) {
        return memo[key];
    }

    long long result;

    // Rule 1: 0 becomes 1
    if (value == 0) {
        result = count_stones(1, blinks - 1);
    }
    // Rule 2: Even number of digits -> split
    else {
        int num_digits = count_digits(value);
        if (num_digits % 2 == 0) {
            auto [left, right] = split_number(value, num_digits);
            result = count_stones(left, blinks - 1) + count_stones(right, blinks - 1);
        }
        // Rule 3: Multiply by 2024
        else {
            result = count_stones(value * 2024, blinks - 1);
        }
    }

    // Store in cache
    memo[key] = result;
    return result;
}

long long part1(const vector<long long>& stones) {
    long long total = 0;
    for (long long stone : stones) {
        total += count_stones(stone, 25);
    }
    return total;
}

long long part2(const vector<long long>& stones) {
    long long total = 0;
    for (long long stone : stones) {
        total += count_stones(stone, 75);
    }
    return total;
}

int main() {
    // Read input file
    ifstream infile("../input.txt");
    if (!infile) {
        cerr << "Error: Could not open input.txt" << endl;
        return 1;
    }

    string line;
    getline(infile, line);
    infile.close();

    // Parse space-separated numbers
    vector<long long> stones;
    stringstream ss(line);
    long long num;
    while (ss >> num) {
        stones.push_back(num);
    }

    // Solve both parts
    cout << "Part 1: " << part1(stones) << endl;

    // Clear memo between parts to avoid any issues
    // (though it shouldn't matter since we're using different blink counts)

    cout << "Part 2: " << part2(stones) << endl;

    return 0;
}

File: 2024/day11/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use v5.10;

# Read input file
my $input_file = '../input.txt';
open my $fh, '<', $input_file or die "Cannot open $input_file: $!";
my $input = do { local $/; <$fh> };
close $fh;

# Parse space-separated numbers
chomp $input;
my @stones = split /\s+/, $input;

# Memoization hash for count_stones function
my %memo;

sub count_stones {
    my ($value, $blinks) = @_;

    # Base case: no more blinks
    return 1 if $blinks == 0;

    # Check memoization cache
    my $key = "$value,$blinks";
    return $memo{$key} if exists $memo{$key};

    my $result;

    # Rule 1: 0 becomes 1
    if ($value == 0) {
        $result = count_stones(1, $blinks - 1);
    }
    # Rule 2: Even number of digits -> split
    elsif (length($value) % 2 == 0) {
        my $s = "$value";
        my $mid = length($s) / 2;
        my $left = int(substr($s, 0, $mid));
        my $right = int(substr($s, $mid));
        $result = count_stones($left, $blinks - 1) + count_stones($right, $blinks - 1);
    }
    # Rule 3: Multiply by 2024
    else {
        $result = count_stones($value * 2024, $blinks - 1);
    }

    # Store in cache
    $memo{$key} = $result;
    return $result;
}

sub part1 {
    my $total = 0;
    for my $stone (@stones) {
        $total += count_stones($stone, 25);
    }
    return $total;
}

sub part2 {
    my $total = 0;
    for my $stone (@stones) {
        $total += count_stones($stone, 75);
    }
    return $total;
}

# Run both parts
say "Part 1: " . part1();
say "Part 2: " . part2();
